libs/vkd3d-shader/spirv.c

```diff
@@ -3199,6 +3185,73 @@ static uint32_t vkd3d_dxbc_compiler_emit_load_scalar(struct vkd3d_dxbc_compiler
     return val_id;
 }
 
+static uint32_t vkd3d_dxbc_compiler_emit_load_constant_buffer(struct vkd3d_dxbc_compiler *compiler,
+        const struct vkd3d_shader_register *reg, const struct vkd3d_shader_register_info *register_info,
+        DWORD swizzle, DWORD write_mask)
+{
+    struct vkd3d_spirv_builder *builder = &compiler->spirv_builder;
+    uint32_t type_id, ptr_type_id, ptr_id, val_id, last_index;
+    uint32_t indexes[4], component_ids[4];
+    unsigned int i, j, component_count;
+
+    assert(!reg->idx[0].rel_addr);
+
+    last_index = 0;
+
+    if (register_info->cbv_binding)
+    {
+        if (register_info->cbv_binding->flags & VKD3D_SHADER_BINDING_FLAG_BINDLESS)
+            indexes[last_index++] = vkd3d_dxbc_compiler_get_resource_index(compiler, reg, register_info->cbv_binding);
+
+        indexes[last_index++] = vkd3d_dxbc_compiler_get_constant_uint(compiler, register_info->member_idx);
+        indexes[last_index++] = vkd3d_dxbc_compiler_emit_register_addressing(compiler,
+                &reg->idx[shader_is_sm_5_1(compiler) ? 2 : 1]);
+    }
+
+    type_id = vkd3d_spirv_get_type_id(builder, register_info->component_type, 1);
+    ptr_type_id = vkd3d_spirv_get_op_type_pointer(builder, register_info->storage_class, type_id);
+
+    for (i = 0, j = 0; i < VKD3D_VEC4_SIZE; i++)
+    {
+        if (!(write_mask & (1 << i)))
+            continue;
+
+        if (register_info->cbv_binding)
+        {
+            indexes[last_index] = vkd3d_dxbc_compiler_get_constant_uint(compiler, vkd3d_swizzle_get_component(swizzle, i));
+        }
+        else
+        {
+            /* Root constants are unrolled to individual floats */
+            indexes[last_index] = vkd3d_dxbc_compiler_get_constant_uint(compiler,
```

Comment:
Since we're unrolling members, we probably need to clamp the member index here for robustness?

Reply:
Hm, should we maybe return `OpUndef` if the index is out of bounds?

Reply:
Either that, or 0. Not sure how OOB for root constants are specified. But the most common case here is when a CBV only contains 8 bytes and app only accesses 8 bytes, but DXBC still deals in terms of 16 byte loads.